runtime/rococo/src/propose_parachain.rs

```diff
@@ -0,0 +1,836 @@
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// The name of the parachain is too long.
+		NameTooLong,
+		/// The requested parachain id is already registered.
+		ParachainIdAlreadyTaken,
+		/// The requested parachain id is already proposed for another parachain.
+		ParachainIdAlreadyProposed,
+		/// Could not find the parachain proposal.
+		ProposalNotFound,
+		/// Not authorized to do a certain operation.
+		NotAuthorized,
+		/// A validator is already registered in the active validator set.
+		ValidatorAlreadyRegistered,
+		/// No information about the registered parachain found.
+		ParachainInfoNotFound,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as ParachainProposer {
+		/// All the proposals.
+		Proposals: map hasher(twox_64_concat) ParaId => Option<Proposal<T::AccountId, T::ValidatorId, BalanceOf<T>>>;
+		/// Proposals that are approved.
+		ApprovedProposals: Vec<ParaId>;
+		/// Proposals that are scheduled at for a fixed session to be applied.
+		ScheduledProposals: map hasher(twox_64_concat) SessionIndex => Vec<ParaId>;
+		/// Information about the registered parachains.
+		ParachainInfo: map hasher(twox_64_concat) ParaId => Option<RegisteredParachainInfo<T::AccountId, T::ValidatorId>>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin, system = system {
+		type Error = Error<T>;
+
+		/// The maximum name length of a parachain.
+		const MaxNameLength: u32 = T::MaxNameLength::get();
+
+		fn deposit_event() = default;
+
+		#[weight = T::MaximumBlockWeight::get()]
+		fn propose_parachain(
+			origin,
+			para_id: ParaId,
+			name: Vec<u8>,
+			validation_function: ValidationCode,
+			initial_head_state: HeadData,
+			validators: [T::ValidatorId; 2],
+			balance: BalanceOf<T>,
+		) {
+			let who = ensure_signed(origin)?;
+
+			ensure!(name.len() <= T::MaxNameLength::get() as usize, Error::<T>::NameTooLong);
+			ensure!(!Proposals::<T>::contains_key(&para_id), Error::<T>::ParachainIdAlreadyProposed);
+			ensure!(!runtime_common::parachains::Code::contains_key(&para_id), Error::<T>::ParachainIdAlreadyTaken);
+
+			let active_validators = Session::<T>::validators();
+			ensure!(
+				active_validators.iter().all(|v| *v != validators[0] && *v != validators[1]),
+				Error::<T>::ValidatorAlreadyRegistered,
+			);
+			Proposals::<T>::iter().try_for_each(|(_, prop)|
+				if prop.validators.iter().all(|v| *v != validators[0] && *v != validators[1]) {
+					Ok(())
+				} else {
+					Err(Error::<T>::ValidatorAlreadyRegistered)
+				}
+			)?;
+
+			T::Currency::reserve(&who, T::ProposeDeposit::get())?;
+
+			let proposal = Proposal {
+				name: name.clone(),
+				proposer: who,
+				validators: validators.into(),
+				initial_head_state,
+				validation_function,
+				balance,
+			};
+
+			Proposals::<T>::insert(para_id, proposal);
+
+			Self::deposit_event(Event::ParachainProposed(name, para_id));
+		}
+
+		#[weight = 100_000]
+		fn approve_proposal(
```

Comment:
This function doesnt do ANY verification that a proposal with this `para_id` exists. Looks like it would be a simple enough sanity check